server/src/unifyfs_request_manager.c

```diff
@@ -1273,6 +1272,27 @@ static int rm_process_client_requests(reqmgr_thrd_t* reqmgr)
     return ret;
 }
 
+/* insert failed client in cleanup_arraylist[] */
+// QUESTION: Pass app & client id's, or just reqmgr and init id's here?
+// QUESTION: Locks here and/or in process loop for main thread?
+void queue_cleanup(reqmgr_thrd_t* thrd_ctrl)
```

Comment:
There are a couple problems here. 

First, you reset `cleanup_arraylist` to NULL, then try to insert things into it, which will fail. I also do not see anywhere in the code where you create the arraylist.

Second, you need to add a single entry to the arraylist that contains the app_id AND client_id. My suggestion would be to lookup the app_client* for the client using `get_app_client()`, and store that pointer in the arraylist.